### PR TITLE
http: add StringMatcher in HeaderMatcher and deprecate the old fields (exact, prefix, etc.)

### PR DESCRIPTION
**Pull request by yangminzhu**
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

Commit Message: Add StringMatcher in HeaderMatcher and deprecate the old fields (exact, prefix, etc.)
Additional Description:
Risk Level: Low
Testing: Unit Tests
Docs Changes:
Release Notes: Added notes for the new field `string_match` and the deprecation of the old fields (exact, prefix, etc.).
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yangminzhu** [comment]
Could someone take a look at this? The `envoy-presubmit` failed seems not related to the PR, thanks.

File: test/common/http/header_utility_test.cc
Comment (soulxu) [inline test/common/http/header_utility_test.cc:421]: also can test a case like 'Match-Value'

File: test/common/http/header_utility_test.cc
Comment (soulxu) [inline test/common/http/header_utility_test.cc:426]: Also can have a test case for 'Match-Value'

**Review by soulxu — COMMENTED** [review]
@yangminzhu the `envoy-presubmit` doesn't seems your fault. :)

File: source/common/http/header_utility.cc
Comment (yangminzhu) [reply source/common/http/header_utility.cc:177]: It's to cache the lower case result to make it a bit faster so that we do not call AsciiStrToLower() on every requests.

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc]: done.

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc]: done.

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:1933]: Nit `*exact_match*`, `*safe_regex*` etc.

File: source/common/http/header_utility.cc
Comment (htuch) [reply source/common/http/header_utility.cc:177]: Why not just use the `value_` field for this?

File: test/common/http/header_utility_test.cc
Comment (htuch) [inline test/common/http/header_utility_test.cc:755]: Why is this a `for` loop? Here and other cases.

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks, seems reasonable, just some minor comments.
/wait

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:1934]: Why does this not apply to `safe_regex_match` as well?

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:1935]: Instead of adding this here, I think it would make more sense to add a [`type.matcher.v3.StringMatcher`](https://github.com/envoyproxy/envoy/blob/dfef049fe9692ff8188a83bab859b310c212ece1/api/envoy/type/matcher/v3/string.proto#L20) field and deprecate the existing `exact_match`, `safe_regex_match`, `prefix_match`, `suffix_match`, and `contains_match` fields.  That way, we can eventually get away from having duplicate implementations here.

Note that `StringMatcher` already supports an `ignore_case` field.

Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto:1935]: Sure I'm fine with StringMatcher, it is a slightly larger API change and deprecation. Also cc @htuch in case if there are any other comments about this.

File: api/envoy/config/route/v3/route_components.proto
Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto:1934]: I'm following the same [ignore_case](https://github.com/envoyproxy/envoy/blob/dfef049fe9692ff8188a83bab859b310c212ece1/api/envoy/type/matcher/v3/string.proto#L65) in StringMatcher which also does not apply to regex, I think probably the reason is you can always enable the case-insensitive match easily with the `(?i)` flag directly in the regular expression itself?

Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto:1934]: Oh, interesting -- I didn't realize that.  And it looks like gRPC C-core has the wrong behavior here.  I'll fix that separately.

File: api/envoy/config/route/v3/route_components.proto
Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto]: updated to use `StringMatcher string_match` instead.

File: source/common/http/header_utility.cc
Comment (yangminzhu) [reply source/common/http/header_utility.cc]: changed to StringMatcher now, this is no longer needed.

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc]: not needed now.

**Comment by yangminzhu** [comment]
@htuch @markdroth updated to use StringMatcher now, PTAL, thanks.

File: api/envoy/config/route/v3/route_components.proto
Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto]: If you think it would be useful, I think we can support it in StringMatcher in a follow PR later.

File: api/envoy/type/matcher/v3/string.proto
Comment (yangminzhu) [inline api/envoy/type/matcher/v3/string.proto:65]: note: this is already supported but just not documented in the API.

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto]: No, that's okay.  I've already fixed gRPC C-core to match what Envoy does (grpc/grpc#26567).  The other gRPC languages do the same thing.

**Review by markdroth — COMMENTED** [review]
This approach looks great to me!  I'll let Harvey do the API approval.

File: configs/terminate_http2_post.yaml
Comment (soulxu) [inline configs/terminate_http2_post.yaml:35]: nit, doing a search in the code, still found some example need to update  for example:
https://github.com/envoyproxy/envoy/blob/main/docs/root/configuration/http/http_filters/tap_filter.rst
https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto#L63

Comment (yangminzhu) [reply configs/terminate_http2_post.yaml:35]: I was planning to only change those causing the test failure but it seems we'd better just change all of them.

**Comment by yangminzhu** [comment]
@htuch would you mind to take a look at the PR again? I changed to use StringMatcher directly suggested by Mark and updated existing config and code to use it accordingly, thanks.

File: api/envoy/config/route/v3/route_components.proto
Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto]: I think we discussed this in the past and there was a concern we might break a lot of exiting users via a deprecation here. OTOH, since we don't have major versions coming up, I think we need some way to pay down API tech debt. I'm good with this if other API shepherds agree, but would like to check with @mattklein123 who had the reservation last time this came up.

Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto]: Yeah I agree this is going to be a very large change for everyone, and kind of annoying, but as @htuch says if v4 is "never" we need to go ahead and do stuff like this. Given that now we are talking about 2 year+ cycles for deprecation removal I think it's fine to go ahead and do it.

Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto]: @htuch @mattklein123 thanks for the comment, the PR is ready for review, PTAL.
I updated again and all tests are passing except the Mac CI test timed out after 3 hours which seems not related to the PR.

File: test/tools/router_check/router.cc
Comment (htuch) [inline test/tools/router_check/router.cc:494]: Nit: not a huge fan of `goto` in C++

File: test/common/http/header_utility_test.cc
Comment (htuch) [inline test/common/http/header_utility_test.cc:292]: Why aren't some of the tests in this file now not `DEPRECATED_FEATURE_TEST`?

**Review by htuch — COMMENTED** [review]
LGTM, just two small things.

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc:292]: Updated to use `DEPRECATED_FEATURE_TEST`.

File: test/tools/router_check/router.cc
Comment (yangminzhu) [reply test/tools/router_check/router.cc]: Updated to remove `FALLTHRU` instead.

File: test/common/http/header_utility_test.cc
Comment (htuch) [inline test/common/http/header_utility_test.cc:204]: I guess I'm still vaguely curious why this was passing CI (with compile_time_options rejecting deprecated fields)?

File: test/common/http/header_utility_test.cc
Comment (htuch) [inline test/common/http/header_utility_test.cc:593]: Do we have sufficient coverage such that when these are removed, the non-`DEPRECATED_FEATURE_TEST` fully cover the alternative `string_match` behaviors?

**Review by htuch — COMMENTED** [review]
LGTM, just some small test questions.
/wait

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc:593]: we should have sufficient coverage, my last commit only added `DEPRECATED_FEATURE_TEST` to the unit tests that are **directly** testing the deprecated fields (i.e. `exact_match`, `prefix_match`, etc). I also added the test for the `string_match` field.

That's also why I didn't add the tag to `MayMatchOneOrMoreRequestHeader` but instead changed the test to use `string_match` field because it's testing more than the deprecated fields.

Comment (htuch) [reply test/common/http/header_utility_test.cc:593]: Makes sense, thanks.

File: test/common/http/header_utility_test.cc
Comment (yangminzhu) [reply test/common/http/header_utility_test.cc:204]: I'm not sure, by looking at the code, `parseHeaderMatcherFromYaml()` calls the `TestUtility::loadFromYaml(yaml, header_matcher)` which seems essentially called the [MessageUtil::loadFromYaml()](https://github.com/envoyproxy/envoy/blob/fa754faf339d9d9180934af710ede614355b48b3/source/common/protobuf/utility.cc#L403-L413) that does not check for the deprecated fields?

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:1944]: @snowp is copying the `StringMatcher` proto to the xds repo in cncf/xds#8.  It might be a good idea to wait for that to land and then use the new type here, since that will save us the trouble of migrating this later.

Comment (yangminzhu) [reply api/envoy/config/route/v3/route_components.proto:1952]: the current StringMatcher is used everywhere in many places, is it better to wait and change all usages at once? It is already a large change for control plane to use the StringMatcher to replace the deprecated fields, just want to make it less complicated by not depending on the extra xds repo for now.

**Comment by yangminzhu** [comment]
I'm not sure why the following tests only failed in CI but passed in my local run, @htuch  @mattklein123  @qiwzhang do you might have any ideas about this? I checked the jwt filter code but has no idea how is it even affected by this PR. thanks.

```
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream
[2021-07-09 23:02:04.849][15090][critical][assert] [test/integration/http_integration.cc:470] assert failure: 0. Details: Timed out waiting for new connection.
```

```
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream
[       OK ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream (68 ms)
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_Http2Downstream_HttpUpstream
[2021-07-09 23:02:22.138][16531][critical][assert] [test/integration/http_integration.cc:470] assert failure: 0. Details: Timed out waiting for new connection.
```

**Comment by htuch** [comment]
@yangminzhu not sure, but with integration tests, machine load and timing can have an impact on flaky tests. Try the advice at https://github.com/envoyproxy/envoy/blob/main/test/integration/README.md#debugging-integration-tests.

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto:1952]: I believe that @snowp was going to change the existing code that uses `StringMatcher` to be templated so that it would work with either copy of `StringMatcher`.  Once that's done, I think it should be trivial to use the new type here.

I think this just comes down to a question of timing.  Snow, how soon do you think you can land the `StringMatcher` change, including any necessary templatizing of code in Envoy?

Comment (snowp) [reply api/envoy/config/route/v3/route_components.proto:1952]: I'm busy the next two days for company perf season, but I'll have more time to work on it starting Wednesday. Early review of https://github.com/envoyproxy/envoy/pull/17096 could be helpful in speeding this up to make sure that there is agreement on the direction.

**Comment by yangminzhu** [comment]
> @yangminzhu not sure, but with integration tests, machine load and timing can have an impact on flaky tests. Try the advice at https://github.com/envoyproxy/envoy/blob/main/test/integration/README.md#debugging-integration-tests.

Thanks for the suggestion. I reproduced the test flaky with `--runs_per_test=1000` in a clean repo at `main` branch which makes me think this PR is unrelated to the test failure. I created https://github.com/envoyproxy/envoy/issues/17303 for tracking the issue.

@mattklein123 @htuch Could you approve the PR if there are no other comments so that we can get this merged while @qiwzhang is working on the test flaky? thanks.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
